File: src/components/lineComponent.js
// @flow

import LineStates from './lineStates';
import Bitmap from '../view/bitmap';
import Line from '../tools/line';
import PointBuilder from '../math/pointBuilder';
import Point from '../math/point';
import Segment from '../math/segment';

export default class LineComponent {
  state: string;
  startPoint: Point;
  lineTool: Line;

  constructor(state: string, lineTool: Line) {
    this.state = state;
    this.startPoint = PointBuilder.defaultPoint();
    this.lineTool = lineTool;
  }

  sendAction(point: Point, bitmap: Bitmap) {
    if (this.state === LineStates.setStartPoint) {
      this.setStartPoint = point;
      this.state = LineStates.setEndPoint;
    } else if (this.state === LineStates.setEndPoint) {
      this.lineTool.draw(bitmap, new Segment(this.startPoint, point));
      this.startPoint = PointBuilder.defaultPoint();
      this.state = LineStates.setStartPoint;
    }
  }

  set setLineTool(value: Line) {
    this.lineTool = value;
  }

  disable() {
    this.state = LineStates.disable;
  }

  enable() {
    this.state = LineStates.setStartPoint;
  }

  set setStartPoint(value: Point) {
    this.startPoint = value;
  }

  get getState(): string {
    return this.state;
  }
}

File: lib/components/line_component.dart
import '../view/bitmap.dart';
import '../view/layer.dart';
import '../tools/line.dart';
import '../math/point.dart';
import '../math/segment.dart';
import 'component.dart';

import 'dart:html';

enum State { SetStartPoint, SetEndPoint }

class LineComponent extends Component {
  State _state;
  Point2D _point;
  Line _tool;

  LineComponent(Line tool) : super(null) {
    _state = State.SetStartPoint;
    _point = Point2DBuilder.defaultPoint();
    _tool = tool;
  }

  void addEvents() {
    if (!hasLayer()) return;
    layer.body.canvas.addEventListener('click', click);
    layer.body.canvas.addEventListener('mousemove', move);
  }

  void removeEvents() {
    if (!hasLayer()) return;
    layer.body.canvas.removeEventListener('click', click);
    layer.body.canvas.removeEventListener('mousemove', move);
  }

  void click(Event event) {
    window.requestAnimationFrame((time) {
      MouseEvent mouseEvent = event as MouseEvent;
      if (state == State.SetStartPoint) {
        _point = _computePoint(mouseEvent, layer.body);
        _state = State.SetEndPoint;
      } else {
        Point2D endPoint = _computePoint(mouseEvent, layer.body);
        tool.draw(layer.body, new Segment(point, endPoint));
        layer.preview.clear();
        _state = State.SetStartPoint;
      }
    });
  }

  void move(Event event) {
    window.requestAnimationFrame((time) {
      if (_state != State.SetEndPoint) return;
      MouseEvent mouseEvent = event as MouseEvent;
      layer.preview.clear();
      Point2D endPoint = _computePoint(mouseEvent, layer.preview);
      tool.draw(layer.preview, new Segment(_point, endPoint));
    });
  }

  void unsetLayer() {
    removeEvents();
    layer = null;
  }

  Point2D _computePoint(MouseEvent event, Bitmap bitmap) {
    Rectangle<num> border = bitmap.canvas.getBoundingClientRect();
    return new Point2D(
        event.client.x - border.left, event.client.y - border.top);
  }

  bool hasLayer() => layer != null;

  State get state => _state;

  Point2D get point => _point;

  Line get tool => _tool;
  void set(Line tool) => _tool = tool;
}
